### ISSUE TITLE
Inconsistent number of references if update job is ran for multiple tags at once

### ISSUE THREAD

**Issue by fstachura** [bootlin/elixir#292]
To reproduce:
1. Pull a git repository with more than one tag
2. Run an update job twice, on an empty data directory each time
3. Run [this](https://gist.github.com/fstachura/ce4babfae2988edce466c69c073e268a) script on references.db from both data directories, compare results between databases. The number of references for some identifiers should be different. 
4. Run [this](https://gist.github.com/fstachura/4b5a0ee1ed0a929d9c967c0f605aaf6d) script on both databases with one of identifiers that has a count difference and compare results (sort before diff). Entries for some files should be missing from one of the databases.

**Comment by tleb** [comment]
The main reason I can see is this timeline:

 - Start indexing of defs in version N;
 - Done indexing of defs in version N;
 - Start indexing of defs in version N+1;
 - Start indexing of refs in version N.

For refs in version N, we take the tokenizer output and tokens that match a def in the database are considered to be refs. We might be catching some defs from version N+1, this depends on timing.

Fixing this would require to have a list of defs in version N available when parsing refs of version N. Those are all defs that come from blobs in version N. This can be expensive to compute. Note that it is not a list of *new* blobs in version N (those that we just parsed), it can be some blobs that have been parsed since a long time that are still part of version N.

**Comment by fstachura** [comment]
> Fixing this would require to have a list of defs in version N available when parsing refs of version N. Those are all defs that come from blobs in version N. This can be expensive to compute. Note that it is not a list of new blobs in version N (those that we just parsed), it can be some blobs that have been parsed since a long time that are still part of version N.

I think it we could just wait for all defs jobs to finish before starting other jobs. I did a simple modification that waits for defs threads to finish before starting everything else and it seems to get consistent results (three attempts on musl with three tags, I tried the default version too and it got different results each time).

It's also likely that we have a similar problem with comps docs:
https://github.com/bootlin/elixir/blob/master/update.py#L520

**Comment by fstachura** [comment]
I don't understand. You still would have to reindex all versions from the start for the references database to be consistent, no?

**Comment by tleb** [comment]
No, if you just computed defs for version N-1, you reuse that for version N. That is an optimization when indexing many versions in a row.

If you don't have easy access to version N-1 defs, you grab them from the database. That's slow (few seconds, complete database read), but required.
